### PR TITLE
Support voting patchsets with the same topic

### PR DESCRIPTION
**Pull request by mobangjack**
- [ ] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your master branch!
- [ ] Ensure that the pull request title represents the desired changelog entry
- [ ] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by mobangjack** [comment]
Hi @rsandell , would you please take a look at this PR?

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java [hunk 1]
```diff
@@ -412,6 +415,8 @@ public void setValues(JSONObject formData) {
             useRestApi = false;
         }
 
+        voteSameTopic = formData.optBoolean("voteSameTopic", false);
```
Comment (rsandell) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java:418]: why reading/writing the same value twice?

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/gerritnotifier/GerritNotifier.java [hunk 2]
```diff
@@ -68,6 +76,42 @@ public GerritNotifier(IGerritHudsonTriggerConfig config, GerritCmdRunner cmdRunn
...
+     * Notify changes with the same topic.
+     *
+     * @param event the change based event.
+     * @param command the gerrit command.
+     */
+    private void notifySameTopic(ChangeBasedEvent event, String command) {
+        Topic topic = event.getChange().getTopicObject();
+        if (topic != null) {
+            String pattern = "CHANGE,PATCHSET".replace("CHANGE", event.getChange().getNumber()).replace("PATCHSET",
+                    event.getPatchSet().getNumber());
```
Comment (rsandell) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/gerritnotifier/GerritNotifier.java:89]: This looks very convoluted and slow compared to a simple string concatenation;

```suggestion
            String pattern = new StringBuilder(event.getChange().getNumber()).append(',').append(event.getPatchSet().getNumber()).toString();
```

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/gerritnotifier/GerritNotifier.java [hunk 3]
```diff
@@ -68,6 +76,42 @@ public GerritNotifier(IGerritHudsonTriggerConfig config, GerritCmdRunner cmdRunn
...
+                    continue;
+                }
+                Map<Change, PatchSet> changes = topic.getChanges(server.getQueryHandler());
+                for (Map.Entry<Change, PatchSet> entry : changes.entrySet()) {
+                    Change change = entry.getKey();
+                    if (change.equals(event.getChange())) {
+                        continue;
+                    }
+                    PatchSet patchSet = entry.getValue();
+                    String substitution = "CHANGE,PATCHSET".replace("CHANGE", change.getNumber())
```
Comment (rsandell) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/gerritnotifier/GerritNotifier.java:105]: dito

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/gerritnotifier/GerritNotifier.java [hunk 4]
```diff
@@ -68,6 +76,42 @@ public GerritNotifier(IGerritHudsonTriggerConfig config, GerritCmdRunner cmdRunn
...
+                            .replace("PATCHSET", patchSet.getNumber());
+                    String command2 = command.replace(pattern, substitution);
```
[the remark this answers is not in the document]
Comment (rsandell) [reply src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/gerritnotifier/GerritNotifier.java:107]: Or have I misunderstood this code?

**Review by rsandell — CHANGES_REQUESTED** [review]

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java (hunk 1 above)
Comment (mobangjack) [reply src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java]: Sorry, it's a mistake. Fixed.

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/gerritnotifier/GerritNotifier.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (mobangjack) [reply src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/gerritnotifier/GerritNotifier.java:108]: After reading src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/config/Config.java I suppose that whatever the build started/completed command is, there will be a pair of  <CHANGE>,<PATCHSET> in the command, e.g.

gerritVerifiedCmdBuildSuccessful = formData.optString(
                "gerritVerifiedCmdBuildSuccessful",
                "gerrit review <CHANGE>,<PATCHSET> --message 'Build Successful <BUILDS_STATS>' "
                        + "--verified <VERIFIED> --code-review <CODE_REVIEW> --tag " + Constants.TAG_VALUE);
Are there other cases?

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/gerritnotifier/GerritNotifier.java (hunk 2 above)
Comment (mobangjack) [reply src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/gerritnotifier/GerritNotifier.java]: Fixed

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/gerritnotifier/GerritNotifier.java (hunk 3 above)
Comment (mobangjack) [reply src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/gerritnotifier/GerritNotifier.java]: Fixed

**Comment by mobangjack** [comment]
@rsandell Please review

**Comment by mobangjack** [comment]
Hi @rsandell, Would you please review?

**Comment by chenhuahuan** [comment]
Hi  @rsandell , what's the progress on this PR? 
Our team is also looking forward to the same feature that **Gerrit Trigger Voting Same Topic** by plugin  itself after build completed.

**Comment by jgarver** [comment]
Thanks @mobangjack for posting this PR.  It fixes the problem I'm running into currently.  @rsandell Any chance of getting this merged?  Is there another maintainer that could review it?  Thanks!

File: src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/gerritnotifier/GerritNotifier.java [hunk 5]
```diff
@@ -68,6 +76,43 @@ public GerritNotifier(IGerritHudsonTriggerConfig config, GerritCmdRunner cmdRunn
...
+        Topic topic = event.getChange().getTopicObject();
+        if (topic != null) {
+            String pattern = new StringBuilder(event.getChange().getNumber()).append(',')
+                    .append(event.getPatchSet().getNumber()).toString();
+
+            if (!command.contains(pattern)) {
+                logger.error("command {} has no pattern {}", command, pattern);
+                return;
+            }
+            for (GerritServer server : PluginImpl.getServers_()) {
```
Comment (ckreisl) [inline src/main/java/com/sonyericsson/hudson/plugins/gerrit/trigger/gerritnotifier/GerritNotifier.java:95]: Checking all Gerrit servers is not required, right? From my understanding so far the build started event is triggered with the corresponding server information so we just need to check whether the current server config has "isVoteSameTopic()" enabled.

**Comment by ckreisl** [comment]
I'm actively working on this feature right now [https://github.com/ckreisl/gerrit-trigger-plugin/tree/feature-add-vote-patches-same-topic](https://github.com/ckreisl/gerrit-trigger-plugin/tree/feature-add-vote-patches-same-topic). Tried to adapt the code with the review feedback to use `ParameterExpander`. Still need to cleanup some parts of the code and write proper unit tests.

Just wanted to give a first look and ask about the best practice of adding the new changes here. Is it possible to add them here to this PR or should I create a new PR and link the corresponding changes.

**Pull request merged**